File: lib/core/utils/geo_discovery/retry.dart
import 'dart:math' as math;

import '../../../config/geo_discovery_config.dart';
import '../telemetry.dart';
import 'geo_types.dart';
import 'rate_guard.dart';

const _retryableCodes = {
  GeoFailureCode.backendTimeout,
  GeoFailureCode.networkError,
  GeoFailureCode.offline,
  GeoFailureCode.rateLimited,
  GeoFailureCode.unknown,
};

Future<T> withGeoRetry<T>(
  Future<T> Function() operation, {
  required String queryFingerprint,
  required String serviceTypes,
}) async {
  var attempt = 0;
  GeoDiscoveryError? lastError;

  while (attempt <= GeoDiscoveryConfig.retryMaxAttempts) {
    try {
      return await operation();
    } catch (error) {
      final geoError = toGeoDiscoveryError(error);
      lastError = geoError;
      final canRetry =
          geoError.retryable &&
          _retryableCodes.contains(geoError.code) &&
          attempt < GeoDiscoveryConfig.retryMaxAttempts;

      if (!canRetry) {
        throw geoError;
      }

      final configuredDelay =
          geoError.retryAfter ??
          GeoDiscoveryConfig.retryDelays[math.min(
            attempt,
            GeoDiscoveryConfig.retryDelays.length - 1,
          )];
      final capped = configuredDelay > GeoDiscoveryConfig.retryMaxDelay
          ? GeoDiscoveryConfig.retryMaxDelay
          : configuredDelay;
      final jitterMs = math.Random().nextInt(
        math.max(capped.inMilliseconds, 1),
      );
      final delay = Duration(milliseconds: jitterMs);

      telemetry.warn(TelemetryEvent.geoSearchFailed, {
        'code': geoError.code.apiValue,
        'durationMs': delay.inMilliseconds,
        'queryFingerprint': queryFingerprint,
        'retryCount': attempt + 1,
        'serviceTypes': serviceTypes,
        'status': 'retrying',
      });
      await Future<void>.delayed(delay);
      attempt += 1;
    }
  }

  throw lastError ??
      const GeoDiscoveryError(
        'Geo discovery retry failed.',
        code: GeoFailureCode.unknown,
      );
}

File: src/core/geo/retry.ts
import { geoDiscoveryConfig } from "@/config/geoDiscovery";
import { GeoDiscoveryError, toGeoDiscoveryError } from "@/core/geo/geoError";
import { sleepWithAbort } from "@/core/geo/rateGuard";
import { geoTelemetry } from "@/utils/telemetry/geoTelemetry";

const retryableCodes = new Set(["backend_timeout", "network_error", "offline", "rate_limited", "unknown"]);

const jitter = (delayMs: number) => Math.round(Math.random() * delayMs);

export const withGeoRetry = async <TValue>(
  operation: () => Promise<TValue>,
  options: {
    queryFingerprint: string;
    serviceTypes: string;
    signal?: AbortSignal;
  },
): Promise<TValue> => {
  let attempt = 0;
  let lastError: GeoDiscoveryError | null = null;

  while (attempt <= geoDiscoveryConfig.rateLimit.retryMaxAttempts) {
    try {
      return await operation();
    } catch (error) {
      const geoError = toGeoDiscoveryError(error);
      lastError = geoError;

      if (
        options.signal?.aborted ||
        !geoError.retryable ||
        !retryableCodes.has(geoError.code) ||
        attempt >= geoDiscoveryConfig.rateLimit.retryMaxAttempts
      ) {
        throw geoError;
      }

      const configuredDelay =
        geoError.retryAfterMs ??
        geoDiscoveryConfig.rateLimit.retryDelaysMs[
          Math.min(attempt, geoDiscoveryConfig.rateLimit.retryDelaysMs.length - 1)
        ] ??
        geoDiscoveryConfig.rateLimit.retryMaxDelayMs;
      const delayMs = Math.min(jitter(configuredDelay), geoDiscoveryConfig.rateLimit.retryMaxDelayMs);

      geoTelemetry.warn("geo_search_failed", {
        code: geoError.code,
        durationMs: delayMs,
        queryFingerprint: options.queryFingerprint,
        retryCount: attempt + 1,
        serviceTypes: options.serviceTypes,
        status: "retrying"
      });

      await sleepWithAbort(delayMs, options.signal);
      attempt += 1;
    }
  }

  throw lastError ?? new GeoDiscoveryError("Geo discovery retry failed", { code: "unknown" });
};
